I18n: Prevent Intl.DateTimeFormat crash with invalid locales

akhilsingh-git opened the pull request:
- Add locale validation utilities to prevent crashes when LANG=c is set
- Filter out invalid locales from navigator.languages before creating DateTimeFormat
- Add fallback handling to use browser defaults when all locales are invalid
- Fixes issue #110494 where Grafana crashes with 'RangeError: Incorrect locale information provided'
- Maintains backward compatibility for valid locales

<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

[Add a brief description of what the feature or update does.]

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

akhilsingh-git committed:
fix: prevent Intl.DateTimeFormat crash with invalid locales like 'c'

- Add locale validation utilities to prevent crashes when LANG=c is set
- Filter out invalid locales from navigator.languages before creating DateTimeFormat
- Add fallback handling to use browser defaults when all locales are invalid
- Fixes issue #110494 where Grafana crashes with 'RangeError: Incorrect locale information provided'
- Maintains backward compatibility for valid locales

Signed-off-by: Akhil Singh <singhakhil69@gmail.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=110522) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=110522) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=110522) it.</sub>

eledobleefe commented:
Hi @akhilsingh-git,

First of all, thanks for raising this PR 🙂 
We were taking a look at the related issue recently and were going to take the approach of using 'try - catch' to catch the error, and if there was one, just set the locale to `undefined`, so it can set the best option automatically. Have you tried this previously to the solution you're sharing here? 🤔 

```jsx
let formatter: Intl.DateTimeFormat

try {
  formatter = new Intl.DateTimeFormat(locale, options)
} catch {
  formatter = new Intl.DateTimeFormat(undefined, options)
}
```

hugohaggmark commented on packages/grafana-data/src/datetime/formats.ts:
```diff
@@ -105,7 +105,13 @@ export function localTimeFormat(
 
   // https://momentjs.com/docs/#/displaying/format/
-  const dateTimeFormat = new Intl.DateTimeFormat(locale || undefined, options);
+  let dateTimeFormat: Intl.DateTimeFormat;
+
+  try {
+    dateTimeFormat = new Intl.DateTimeFormat(locale || undefined, options);
+  } catch {
+    dateTimeFormat = new Intl.DateTimeFormat("en-US", options);
+  }
```
This looks great, maybe we could extract this part into its own function that could be reused in  `packages/grafana-i18n/src/dates.ts` and `public/app/core/internationalization/dates.ts` as well. We should probably add some tests for this too 👍

akhilsingh-git replied:
sure @hugohaggmark, i have:
Created a reusable function (packages/grafana-i18n/src/safe-datetime-format.ts)
Added comprehensive tests (packages/grafana-i18n/src/safe-datetime-format.test.ts)
Updated all three files to use the new function (packages/grafana-i18n/src/dates.ts, public/app/core/internationalization/dates.ts, packages/grafana-data/src/datetime/formats.ts)

hugohaggmark replied:
Looks great 🎉 , thank you for taking the time to fix that 🙌  I'm leaving the final approval to @grafana/grafana-frontend-platform because the date formatting is not my expertise 👍

akhilsingh-git committed:
refactor: simplify locale fix to use direct try-catch approach

- Remove locale-utils.ts and locale-utils.test.ts files
- Use simple try-catch in dates.ts files: try locale, fallback to 'en-US'
- Use Laura's suggested approach in formats.ts with explicit variable declaration
- Remove unused utility exports from index.ts
- Maintains same functionality with cleaner, simpler code
- Avoids adding to public API while still preventing crashes with invalid locales like 'c'

Signed-off-by: Akhil Singh <singhakhil69@gmail.com>

akhilsingh-git committed:
style: run prettier to fix linting issues

- Format code according to project prettier configuration
- Fixes failing Lint Frontend check

Signed-off-by: Akhil Singh <singhakhil69@gmail.com>
